shared/chat/conversation/list/index.desktop.js

```diff
@@ -1,244 +1,269 @@
-import shallowEqual from 'shallowequal'
-import {AutoSizer, CellMeasurer, List as VirtualizedList, defaultCellMeasurerCellSizeCache as DefaultCellMeasurerCellSizeCache} from 'react-virtualized'
 import {Icon} from '../../../common-adapters'
 import {TextPopupMenu, AttachmentPopupMenu} from '../messages/popup'
 import {clipboard} from 'electron'
+import {debounce} from 'lodash'
 import {findDOMNode} from '../../../util/dom'
 import {globalColors, globalStyles} from '../../../styles'
 
-import type {List} from 'immutable'
 import type {Props} from '.'
 
-type DefaultCellRangeRendererParams = {
-  cellCache: Object,
-  cellRenderer: Function,
-  columnSizeAndPositionManager: Object,
-  columnStartIndex: number,
-  columnStopIndex: number,
-  horizontalOffsetAdjustment: number,
-  isScrolling: boolean,
-  rowSizeAndPositionManager: Object,
-  rowStartIndex: number,
-  rowStopIndex: number,
-  scrollLeft: number,
-  scrollTop: number,
-  styleCache: Object,
-  verticalOffsetAdjustment: number,
-  visibleColumnIndices: Object,
-  visibleRowIndices: Object,
-}
-
 type State = {
   isLockedToBottom: boolean,
-  isScrolling: boolean,
-  messages: List<Constants.Message>,
-  scrollTop: number,
-  selectedMessageID?: Constants.MessageID,
+  listRerender: number,
+  selectedMessageKey: ?Constants.MessageKey,
 }
 
-const scrollbarWidth = 20
 const lockedToBottomSlop = 20
 const listBottomMargin = 10
-const DEBUG_ROW_RENDER = __DEV__ && false
-
-class ConversationList extends Component<void, Props, State> {
-  _cellCache: any;
-  _cellMeasurer: any;
-  _list: any;
-  state: State;
-  _toRemeasure: Array<number>;
-  _shouldForceUpdateGrid: boolean;
-  _lastWidth: ?number;
-
-  constructor (props: Props) {
-    super(props)
-
-    this.state = {
-      isLockedToBottom: true,
-      isScrolling: false,
-      messages: props.messages,
-      scrollTop: 0,
-    }
 
-    this._cellCache = new DefaultCellMeasurerCellSizeCache({
-      uniformColumnWidth: true,
-      uniformRowHeight: false,
-    })
-    this._toRemeasure = []
-    this._shouldForceUpdateGrid = false
+class BaseList extends Component<void, Props, State> {
+  _cellCache = new Virtualized.CellMeasurerCache({
+    fixedWidth: true,
+    keyMapper: (rowIndex: number) => this.props.messageKeys.get(rowIndex),
+  })
 
-    this._setupDebug()
-  }
+  _list: any
+  _keepIdxVisible: number = -1
+  _lastRowIdx: number = -1
 
-  _setupDebug () {
-    if (__DEV__ && typeof window !== 'undefined') {
-      window.dumpChat = (...columns) => {
-        console.table(this.state.messages.toJS().map(m => ({
-          ...m,
-          decoded: m.message && m.message.stringValue(),
-        })), columns.length && columns)
-      }
-    }
-  }
+  // This is similar to a state but it changing value doesn't actually need to cause us to re-render.
+  // We're either locked at the bottom or not and it only affects when we get new messages which'll cause a render anyways
+  _scrollTop = 0
```

Comment:
🔥  leftover from me trying to keep the old scrolltop behavior. i mentioned this on slack. we used to keep actual track of scrolltop but in practice its not perfect. if you scroll up your position isn't kept perfectly (not a big deal right now) with this we try and use scrollToItem to keep your position. its also not perfect but it reduces how much we cause the list to re-render. more details below